shuf: treat -e as a flag, not as a multi-value arg

BenWiederhake opened the pull request:
This PR implements and tests support for the weird permitted argument order of `shuf`, especially for invocations like `shuf -ez firstarg secondarg`, where the `z` shall not be interpreted as a value to `-e`.

This moves us closer to passing the GNU testsuite.

Related:
- Discussed in https://github.com/uutils/coreutils/issues/4254#issuecomment-1953201152 (turns out, I was wrong, it *is* possible)
- Relates to https://github.com/uutils/uutils-args/pull/104 (which is kinda the same patch, but using the new `uutils-arg`).

This PR sadly slightly degrades the quality of the error message in one case:

Before:

```console
$ cargo run -- a b
error: unexpected argument 'b' found

Usage: /path/to/target/debug/shuf [OPTION]... [FILE]
       /path/to/target/debug/shuf -e [OPTION]... [ARG]...
       /path/to/target/debug/shuf -i LO-HI [OPTION]...

For more information, try '--help'.
[$? = 1]
```

After:

```console
$ cargo run -- a b
/path/to/target/debug/shuf: unexpected argument 'b' found
Try '/path/to/target/debug/shuf --help' for more information.
[$? = 1]
```

Note that this does not affect the `--help` message, so I think it's good enough.

tertsdiepraam commented on src/uu/shuf/src/shuf.rs:
```diff
@@ -59,13 +62,19 @@ pub fn uumain(args: impl uucore::Args) -> UResult<()> {
-                .unwrap_or("-")
-                .to_string(),
-        )
+        match matches.get_many::<String>(options::FILE_OR_ARGS) {
+            Some(mut args) if args.len() == 1 => Mode::Default(args.next().unwrap().to_string()),
+            Some(mut args) => {
+                debug_assert!(args.len() >= 2);
+                args.next().unwrap();
+                let second_file = args.next().unwrap();
+                return Err(UUsageError::new(
+                    1,
+                    format!("unexpected argument '{second_file}' found"),
+                ));
+            }
+            None => Mode::Default("-".to_string()),
+        }
```
I've tried to reduce the branching and unwraps a bit. Let me know what you think, it's a tricky bit of code.
```suggestion
        let mut operands = matches
            .get_many::<String>(options::FILE_OR_ARGS)
            .unwrap_or_default();

        let file = operands
            .next()
            .map(Clone::clone)
            .unwrap_or("-".into())
            .to_string();

        if let Some(second_file) = operands.next() {
            return Err(UUsageError::new(
                1,
                format!("unexpected argument '{second_file}' found"),
            ));
        };

        Mode::Default(file)
```

tertsdiepraam commented on src/uu/shuf/src/shuf.rs:
```diff
@@ -42,15 +42,18 @@ mod options {
-    let mode = if let Some(args) = matches.get_many::<String>(options::ECHO) {
-        Mode::Echo(args.map(String::from).collect())
+    let mode = if matches.get_flag(options::ECHO) {
+        match matches.get_many::<String>(options::FILE_OR_ARGS) {
+            Some(args) => Mode::Echo(args.map(String::from).collect()),
+            None => Mode::Echo(vec![]),
+        }
```
I think this is slightly nicer, not by much though.

```suggestion
        Mode::Echo(
            matches
                .get_many::<String>(options::FILE_OR_ARGS)
                .unwrap_or_default()
                .map(String::from)
                .collect(),
        )
```

tertsdiepraam reviewed (commented):
Great! Thanks!

BenWiederhake commented:
Changes since last push:
- Rebased on newest main
- Applied two cleanups – thanks @tertsdiepraam!

BenWiederhake commented:
Changes since last push:
- Fix two lints that aren't in pre-commit, but were caught by CI (specifically: prefer `.cloned()`, and drop an unnecessary `.to_string()`)

cakebaker commented on src/uu/shuf/src/shuf.rs:
```diff
@@ -59,13 +65,17 @@ pub fn uumain(args: impl uucore::Args) -> UResult<()> {
+            .get_many::<String>(options::FILE_OR_ARGS)
+            .unwrap_or_default();
+        let file = operands.next().cloned().unwrap_or("-".into());
+        if let Some(second_file) = operands.next() {
+            return Err(UUsageError::new(
+                1,
+                format!("unexpected argument '{second_file}' found"),
+            ));
+        };
```
Can you please add a test for this error case?

The other changes look good :)

BenWiederhake commented:
Changes since last push:
- Rebased on current main
- Added two new tests for error inputs, specifically multiple files (`test_shuf_two_input_files` and `test_shuf_three_input_files`). Thanks @cakebaker!
- Added a new test for a single dash as filename. (`test_explicit_stdin_file`)

BenWiederhake commented:
Both failures seem to be flakes:
- `test_split::test_round_robin_limited_file_descriptors` (`split: unable to open 'xav'; aborting`)
- `test_timeout::test_kill_subprocess` (`panicked at ''' does not contain 'xyz''`)

tertsdiepraam commented:
Thanks! Your updates on the force pushes are great by the way. Really useful while reviewing!
